### PR TITLE
Add type hints to `skll.data` module & improve types in `skll.config` 

### PR DESCRIPTION
**Pull request by desilinguist**
1. Update pre-commit checks to include a check for [commit message format](https://www.conventionalcommits.org/en/v1.0.0/#summary) and to exclude missing imports from the `mypy` check. 
2. Add new custom types to `skll.types` module.
3. Add type hints to `skll.data` module
    - Refactor code to ensure correct types.
    - Add a lot of checks for labels, features, vectorizer being ``None`` that was previously unhandled.
    - Update docstrings to use the type hints.
4. Improve type hints in `config/__init__.py`
    - Use new `ClassMap` custom type.
    - Make sure `pos_label` is typed correctly.
    - Fix incorrect default value for sampler parameters.

### CODE REVIEW

File: skll/types.py [hunk 1]
```diff
@@ -4,13 +4,39 @@
...
 
 # a mapping from example ID to fold ID;
 # the example ID may be a float or a str
 FoldMapping = Dict[Union[float, str], str]
+
+# a float or a string; this is useful
+# for SKLL IDs that can be both
+IdType = Union[float, str]
+
+# a float or a string; this is useful
```
Comment (tamarl08) [inline skll/types.py:28]:
```suggestion
# a float, an int, or a string; this is useful
```

File: skll/data/featureset.py [hunk 2]
```diff
@@ -9,37 +9,38 @@
...
+from skll.data.dict_vectorizer import DictVectorizer as SkllDictVectorizer
+from skll.types import FeatGenerator, FeatureDictList, IdType, LabelType, SparseFeatureMatrix
 
 
 class FeatureSet(object):
-
     """
-    Encapsulation of all of the features, values, and metadata about a given
-    set of data. This replaces ``ExamplesTuple`` from older versions of SKLL.
+    Encapsulate features, labels, and medata for a given dataset.
```
Comment (tamarl08) [inline skll/data/featureset.py:25]:
```suggestion
    Encapsulate features, labels, and metadata for a given dataset.
```

File: skll/data/featureset.py [hunk 3]
```diff
@@ -57,26 +58,49 @@ class FeatureSet(object):
...
+        labels: Optional[Union[List[str], np.ndarray]] = None,
+        features: Optional[Union[FeatureDictList, SparseFeatureMatrix]] = None,
+        vectorizer: Optional[Union[DictVectorizer, FeatureHasher]] = None,
+    ):
+        """Initialize a FeatureSet instance."""
         super(FeatureSet, self).__init__()
+
+        # clearly define the attribute types
+        self.ids: np.ndarray
+        self.labels: Optional[np.ndarray]
```
Comment (tamarl08) [inline skll/data/featureset.py:74]: what does Optional mean in the context of defining the attribute type?

File: skll/data/featureset.py [hunk 4]
```diff
@@ -154,16 +175,16 @@ def __iter__(self):
...
+        return self.features.shape[0] if self.features is not None else 0
 
-    def __add__(self, other):
+    def __add__(self, other: "FeatureSet") -> "FeatureSet":
         """
-        Combine two feature sets to create a new one.  This is done assuming
-        they both have the same instances with the same IDs in the same order.
+        Combine two feature sets to create a new one.
+
+        Thie combination is done assuming they both have the same instances
```
Comment (tamarl08) [inline skll/data/featureset.py:186]:
```suggestion
        The combination is done assuming they both have the same instances
```

File: skll/data/featureset.py [hunk 5]
```diff
@@ -345,20 +401,24 @@ def filtered_iter(self, ids=None, labels=None, features=None, inverse=False):
...
-        id_ : str
+        id_ : IdType
             The ID of the example.
 
-        label_ : str
+        label_ : LabelType
             The label of the example.
 
-        feat_dict : dict
+        feat_dict : Dict[str, Any]
```
Comment (tamarl08) [inline skll/data/featureset.py:410]:
```suggestion
        feat_dict : FeatureDict
```

File: skll/data/featureset.py [hunk 6]
```diff
@@ -469,30 +546,34 @@ def __getitem__(self, value):
...
         else:
             label = self.labels[value] if self.labels is not None else None
-            feats = self.features[value, :]
-            features = (
-                self.vectorizer.inverse_transform(feats)[0] if self.features is not None else {}
-            )
+            feats = self.features[value, :] if self.features is not None else None
+            if self.vectorizer:
+                features = (
+                    self.vectorizer.inverse_transform(feats)[0] if self.features is not None else {}
```
Comment (tamarl08) [inline skll/data/featureset.py:552]: what will the return value be if feats is None?

File: skll/data/readers.py [hunk 7]
```diff
@@ -88,10 +92,12 @@ class Reader(object):
         If no column with that name exists, or ``None`` is
         specified, example IDs will be automatically generated.
 
-    class_map : dict, default=None
+    class_map : Optional[Dict[str, List[str]]], default=None
```
Comment (tamarl08) [inline skll/data/readers.py:95]:
```suggestion
    class_map : Optional[ClassMap], default=None
```

File: skll/data/featureset.py (hunk 6 above)
Comment (desilinguist) [reply skll/data/featureset.py:552]: Good catch! Probably `feats` should be an empty numpy array for it go through?

File: skll/data/readers.py [hunk 8]
```diff
@@ -138,13 +144,15 @@ def __init__(
         self._use_pandas = False
 
         if feature_hasher:
-            self.vectorizer = FeatureHasher(n_features=num_features)
+            self.vectorizer: Union[DictVectorizer, FeatureHasher] = FeatureHasher(
```
Comment (tamarl08) [inline skll/data/readers.py:147]: maybe define the type of vectorizer before this if/else block?

File: skll/data/readers.py [hunk 9]
```diff
@@ -266,8 +276,8 @@ def _sub_read_rows(self, file):
             If the example IDs are not unique.
         """
         # Get labels and IDs
-        ids = []
-        labels = []
+        ids_list: List[Union[float, str]] = []
```
Comment (tamarl08) [inline skll/data/readers.py:279]:
```suggestion
        ids_list: List[IdType] = []
```

File: skll/data/readers.py [hunk 10]
```diff
@@ -266,8 +276,8 @@ def _sub_read_rows(self, file):
...
+        labels_list: List[str] = []
```
Comment (tamarl08) [inline skll/data/readers.py:280]: Can it be LabelType or must be a string?

File: skll/data/readers.py [hunk 11]
```diff
@@ -350,7 +368,7 @@ def _parse_dataframe(self, df, id_col, label_col, replace_blanks_with=None, drop
         labels : np.array of shape (n_labels,)
             The labels for the feature set.
 
-        features : list of dicts
+        features : List[Dict[str, Any]]
```
Comment (tamarl08) [inline skll/data/readers.py:371]:
```suggestion
        features : FeatureDictList
```

File: skll/data/readers.py [hunk 12]
```diff
@@ -454,6 +472,9 @@ def read(self):
             print(self._progress_msg, end="\r", file=sys.stderr)
             sys.stderr.flush()
 
+        # if we are in this method, self.path_or_file must be a path
+        assert isinstance(self.path_or_list, (str, Path)), "file path or path object required"
+
```
Comment (tamarl08) [inline skll/data/readers.py:477]: why can't we convert it to a Path?

Comment (tamarl08) [reply skll/data/readers.py:477]: Oh I see it's a dict.

Comment (tamarl08) [reply skll/data/readers.py:477]: usually we prefer raise error over assert, right?

Comment (desilinguist) [reply skll/data/readers.py:477]: Yes, we do. Let me try and refactor this a bit.

File: skll/data/readers.py [hunk 13]
```diff
@@ -590,8 +617,8 @@ def _sub_read(self, file):
             example = json.loads(line)
             # Convert all IDs to strings initially,
             # for consistency with csv formats.
-            curr_id = str(example.get("id", f"EXAMPLE_{example_num}"))
-            class_name = (
+            curr_id: Union[float, str] = str(example.get("id", f"EXAMPLE_{example_num}"))
```
Comment (tamarl08) [inline skll/data/readers.py:620]:
```suggestion
            curr_id: IdType = str(example.get("id", f"EXAMPLE_{example_num}"))
```

File: skll/data/readers.py [hunk 14]
```diff
@@ -1020,13 +973,13 @@ def _sub_read(self, file):
```
Comment (tamarl08) [inline skll/data/readers.py:973]: line 971: file buffer --> IO[str] ?

File: skll/data/writers.py [hunk 15]
```diff
@@ -581,7 +617,7 @@ def _write_header(self, feature_set, output_file, filter_features):
         # Finish header and start data section
         print("\n@data", file=output_file)
 
-    def _write_line(self, id_, label_, feat_dict, output_file):
+    def _write_line(self, id_, label_, feat_dict, output_file) -> None:
```
Comment (tamarl08) [inline skll/data/writers.py:620]: no parameter types here!

File: skll/data/featureset.py [hunk 16]
```diff
@@ -469,30 +546,34 @@ def __getitem__(self, value):
...
-        ids_for_split1 : list of int
+        ids_for_split1 : List[int]
             A list of example IDs which will be split out into
             the first ``FeatureSet`` instance. Note that the
             FeatureSet instance will respect the order of the
             specified IDs.
 
-        ids_for_split2 : list of int, default=None
+        ids_for_split2 : Optional[List[int]], default=None
             An optional ist of example IDs which will be
```
Comment (Frost45) [inline skll/data/featureset.py:577]:
```suggestion
            An optional list of example IDs which will be
```

File: skll/data/featureset.py [hunk 17]
```diff
@@ -196,67 +220,92 @@ def __add__(self, other):
...
 
         Parameters
         ----------
-        ids : list of str/float, default=None
+        ids : Optional[List[FloatOrStr]], default=None
             Examples to keep in the FeatureSet. If ``None``, no ID
             filtering takes place.
 
-        labels : list of str/float, default=None
+        labels : Optional[List[FloatOrStr]], default=None
```
[the remark this answers is not in the document]
Comment (desilinguist) [reply skll/data/featureset.py:304]: Turn out it already was in the function signature, just an oversight in the docstring.

File: skll/data/readers.py (hunk 12 above)
Comment (desilinguist) [reply skll/data/readers.py:477]: Actually, this assert statement will raise an `AssertionError` if it fails so it's the same thing.

File: skll/data/writers.py [hunk 18]
```diff
@@ -581,22 +617,24 @@ def _write_header(self, feature_set, output_file, filter_features):
...
         Write the current line in the file in this Writer's format.
 
         Parameters
         ----------
-        id_ : str
+        id_ : IdType
             The ID for the current instance.
 
-        label_ : str
+        label_ : LabelTyp
```
[the remark this answers is not in the document]
Comment (desilinguist) [reply skll/data/writers.py:631]: 🤦🏽

**Review by tamarl08 — APPROVED** [review]

**Comment by desilinguist** [comment]
This is part of #561.

**Pull request merged**